gadopt/scalar_equation.py

```diff
@@ -18,69 +18,50 @@
 
-from .equations import BaseEquation, BaseTerm
-from .utility import cell_edge_integral_ratio, is_continuous, normal_is_continuous
+from .equations import Equation, interior_penalty_factor
+from .utility import is_continuous, normal_is_continuous
 
 
-class ScalarAdvectionTerm(BaseTerm):
+def scalar_advection_term(
+    eq: Equation, trial: Argument | ufl.indexed.Indexed | Function
+) -> Form:
     r"""Scalar advection term (non-conservative): u \dot \div(q)."""
-    def residual(
-        self,
-        test: fd.ufl_expr.Argument,
-        trial: fd.ufl_expr.Argument | fd.Function,
-        trial_lagged: Optional[fd.ufl_expr.Argument | fd.Function] = None,
-        fields: Optional[dict[str, fd.Constant | fd.Function]] = None,
-        bcs: Optional[dict[int, dict[str, int | float]]] = None,
-    ) -> fd.ufl.core.expr.Expr:
-        u = fields['velocity']
-        phi = test
-        n = self.n
-        q = trial
-        continuous_u_normal = normal_is_continuous(u)
-        if 'advective_velocity_scaling' in fields:
-            u = fields['advective_velocity_scaling'] * u
-
-        if 'su_nubar' in fields:
-            # SU(PG) ala Donea & Huerta 2003
-            nubar = fields['su_nubar']
-            phi = phi + nubar / (dot(u, u)+1e-12) * dot(u, grad(phi))
-
-            F = phi * dot(u, grad(q)) * self.dx  # The advection term is not integrated by parts so there are no boundary terms
-
-        else:
-            F = -q*div(phi*u)*self.dx
-
-            # integration by parts leads to boundary term
-            F += q*dot(n, u)*phi*self.ds
-
-            if not (is_continuous(self.trial_space) and continuous_u_normal):
-                # s=0: u.n(-)<0  =>  flow goes from '+' to '-' => '+' is upwind
-                # s=1: u.n(-)>0  =>  flow goes from '-' to '+' => '-' is upwind
-                s = 0.5*(sign(dot(avg(u), n('-'))) + 1.0)
-                q_up = q('-')*s + q('+')*(1-s)
-                F += jump(phi*u, n) * q_up * self.dS
-
-        # replace boundary value on incoming Dirichlet 'q' boundaries:
-        for id, bc in bcs.items():
-            if 'q' in bc:
-                # on incoming boundaries, dot(u,n)<0, replace q with bc['q']
-                F += phi*min_value(dot(u, n), 0)*(bc['q']-q) * self.ds(id)
-
-        return -F
-
-
-class ScalarDiffusionTerm(BaseTerm):
+    if hasattr(eq, "advective_velocity_scaling"):
+        u = eq.advective_velocity_scaling * eq.u
+    else:
+        u = eq.u
+
+    if hasattr(eq, "su_nubar"):  # SU(PG) à la Donea & Huerta (2003)
+        phi = eq.test + eq.su_nubar / (dot(u, u) + 1e-12) * dot(u, grad(eq.test))
+
+        # The advection term is not integrated by parts so there are no boundary terms
+        F = phi * dot(u, grad(trial)) * eq.dx
+    else:
+        F = -trial * div(eq.test * u) * eq.dx
+        F += trial * dot(eq.n, u) * eq.test * eq.ds  # Boundary term in the weak form
+
+        if not all([is_continuous(eq.trial_space), normal_is_continuous(eq.u)]):
```

Comment:
I prefer the original

```suggestion
        if not (is_continuous(eq.trial_space) and normal_is_continuous(eq.u)):
```

Reply:
@stephankramer @w3168 Any idea if this was intended to be `not x and not y` or `not x and y`? I would tend to say the former, but then it disagrees with what we have in `master`.

Reply:
hmm I wonder if the original logic was actually right... 

I remember years ago finding a bug with @stephankramer where we had copied the code from the DG scalar advection demo in Firedrake and this assumed that the velocity was in a continuous function space. 

Here's the demo: https://www.firedrakeproject.org/demos/DG_advection.py.html

and the commit in thwaties where we changed it: https://github.com/thwaitesproject/thwaites/commit/c291d09c959fd506d639cf05cf709cb7ede2ecb0#diff-8b82ede7584525921942c3de6d02bea35e9a0fd6a14652133e7bce49d488a792L43

So I think the code in Gadopt (and thwaites) now is the right thing to do if both the tracer and the velocity are discontinuous. I think should satisfy the existing if statement.  (@Patol75 I don't quite see what you mean about it disagreeing with what is currently in master?)

For context when doing ocean modelling with thwaites we always used to use P1DG for velocity and tracers (and then later a tensor product element that was discontinuous in the normal direction...).

I wonder if the gadopt usecase with P2 for velocity and DG for the tracers is doing the right thing... It probably should use the example firedrake code (though it might end  end up being the same as the modified thwaites code when the (normal) velocity is continuous). Would it actually be turned on by the current (and thomas') if statement though?

Reply:
Quick answer about the confusion: Python evaluates `not x and y` as `(not x) and (y)`, which is different from `not x and not y`. So if `x` and `y` both evaluate to `False`, then `not x and y` evaluates to False, whereas `not x and not y` evaluates to True. Does that change your answer?

Reply:
Neverming, I misread the parentheses in the code...

Reply:
I'm probably being stupid but isnt the not outside a bracket... so it evaluates not(x and y) as opposed to not(x) and y which you seem to be saying...

I think the original ocean modelling case: 
x: tracer is continuous - > False
y: velocity is continuous -> False
x and y -> False
so not (x and y) -> True, i.e. add the extra ufl in.



but for mantle convection now... 
x: tracer is continous -> False
y: velocity is continuous -> True
x and y = False
so not(x and y) -> True add the extra ufl in.  
(edit: Ah so I was slightly wrong above. The mantle convection case will also add the ufl in ... with the caveat this might be wrong if u is continuous, but it probably is ok... )

Reply:
I guess this logic might catch the case where you have continuous tracers but discontinuous velocity...?